modules/core/src/test/java/org/apache/ignite/internal/processors/compute/IgniteComputeWithPendingTopologyDeadlockTest.java

```diff
@@ -0,0 +1,304 @@
+                                + "tx - " + concurrency + "/" + isolation + " (see thread dumps).",
+                            doTestTx(concurrency, isolation, cacheMode, cacheWriteSyncMode));
+    }
+
+    /**
+     * Test execution within transaction.
+     *
+     * @param concurrency transaction concurrency level.
+     * @param isolation transaction isolation level.
+     * @return {@code True} if no hanging was detected.
+     * @throws Exception if fails.
+     */
+    private boolean doTestTx(
+        final TransactionConcurrency concurrency,
+        final TransactionIsolation isolation,
+        final CacheMode cacheMode,
+        final CacheWriteSynchronizationMode cacheWriteMode) throws Exception {
+
+        try (Ignite node = startGrid(0)) {
+            final IgniteCache<Integer, Integer> cache1 = node.createCache(cacheConfiguration(CACHE1, cacheMode, cacheWriteMode));
+            final IgniteCache<Integer, Integer> cache2 = node.createCache(cacheConfiguration(CACHE2, cacheMode, cacheWriteMode));
+
+            // Fill with test values.
+            fillCaches(ImmutableList.of(cache1, cache2));
+
+            final CountDownLatch sync = new CountDownLatch(1);
+
+            IgniteInternalFuture<Boolean> fut = runAsync(new Callable<Boolean>() {
+                @Override public Boolean call() throws IgniteInterruptedCheckedException {
+
+                    try (Transaction tx = node.transactions().txStart(concurrency, isolation)) {
+
+                        cache1.put(12, 12);
+
+                        // Signal main thread.
+                        sync.countDown();
+
+                        // Await when new cache will be created.
+                        boolean done = GridTestUtils.waitForCondition(new PA() {
+                            @Override public boolean apply() {
+                                return node.cacheNames().size() >= 3;
+                            }
+                        }, CACHE_CREATION_TIMEOUT);
+
+                        assertTrue(done);
+
+                        // Execute distributed (topology version aware) task.
+                        cache2.clear();
+
+                        tx.commit();
+                    }
+                    catch (CacheException ignore) {
+                        // Expected behavior - computation was aborted.
+                    }
+                    return true;
+                }
+            });
+            // Create watch-thread that will cancel task if timeout expired.
+            IgniteInternalFuture<Boolean> trackingFuture = startTracking(fut);
+
+            // Await new thread.
+            sync.await(PENDING_TIMEOUT, TimeUnit.MILLISECONDS);
+
+            // Create cache - topology update.
+            node.createCache(cacheConfiguration(CACHE3, cacheMode, cacheWriteMode));
+
+            // Await completion or timeout.
+            return trackingFuture.get();

```

Comment:
`.get()` method doesn't throw or return on any timeout. One should use `.get(timeout);` method to prevent hanging.